client/network/src/discovery.rs

```diff
@@ -479,6 +573,170 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 	}
 }
 
+pub struct DiscoveryHandler {
+	handlers: Vec<(ProtocolId, KademliaHandler<QueryId>)>
+}
+
+impl ProtocolsHandler for DiscoveryHandler {
+	type InEvent = (ProtocolId, <KademliaHandler<QueryId> as ProtocolsHandler>::InEvent);
+	type OutEvent = (ProtocolId, <KademliaHandler<QueryId> as ProtocolsHandler>::OutEvent);
+	type Error = <KademliaHandler<QueryId> as ProtocolsHandler>::Error;
+	type InboundProtocol = KadUpgrade<<KademliaHandler<QueryId> as ProtocolsHandler>::InboundProtocol>;
+	type OutboundProtocol = <KademliaHandler<QueryId> as ProtocolsHandler>::OutboundProtocol;
+	type OutboundOpenInfo = (ProtocolId, <KademliaHandler<QueryId> as ProtocolsHandler>::OutboundOpenInfo);
+
+	fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol> {
+		let upgrades = self.handlers.iter()
+			.map(|h| (h.0.clone(), h.1.listen_protocol().into_upgrade().1))
+			.collect();
+		SubstreamProtocol::new(KadUpgrade { upgrades })
+	}
+
+	fn inject_fully_negotiated_outbound
+		( &mut self
```

Comment:
As with the previous PR, I'm really not a fan of this code style. This isn't Haskell.